### PR TITLE
Adding unit tests

### PR DESCRIPTION
**Pull request by tymondesigns**
Compliments #133 

So I thought I would go ahead and throw the initial setup together so everyone can see what the code could look like.

All tests are green 😄

### CODE REVIEW

File: test/run.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+var path = require('path');
```
Comment (anselmh) [inline test/run.js]: Can we perhaps enforce strict mode, please, via `'use strict'`?

**Comment by anselmh** [comment]
Looks good to me at a first glance. I think it’s also okay to have this only available on npm since this is only needed for devs who want to contribute to the repo.

This said, we should add unit tests for all our functions at once, to manage 100% test coverage of our internal mixins/functions? At least I’d vote for this option.

**Comment by anselmh** [comment]
> All tests are green 😄

Haha! Indeed we should add a CI for this. I’ll file a ticket.

**Comment by tymondesigns** [comment]
oh yes 😄 .travis.yml would do nicely

just wanted to get the go ahead before going to town on it...

File: test/run.js (hunk 1 above)
Comment (tymondesigns) [inline test/run.js]: done 👍

**Comment by anselmh** [comment]
https://github.com/inuitcss/inuitcss/issues/135

**Comment by csswizardry** [comment]
@tymondesigns: I’ve been looking into Travis as well—I’d love to run some tests against inuitcss to check that our mixins etc. all work, and that we can completely compile the project right out of the box (e.g. https://mattandre.ws/2014/01/testing-sass-travis-ci/)

**Comment by tymondesigns** [comment]
@csswizardry Yea travis is awesome... I always find myself coming back to it after trying out other CI services.

Great article and I idea regarding testing the compilation - I think we would need to add in node-sass cli for that. Depends on how you feel about adding another dev-dependency I guess...

I'll add in a `.travis.yml` anyways

another good one from sitepoint that I stumbled on - https://www.sitepoint.com/creating-tests-sass-framework/

**Comment by tymondesigns** [comment]
guess we should have the CI discussion over on #135

**Comment by nenadjelovac** [comment]
@tymondesigns shouldn't there be a test for `@include inuit-font-size(21px, $important: true);` as well?

**Comment by tymondesigns** [comment]
@nenadjelovac isn't that covered here - https://github.com/inuitcss/inuitcss/pull/134/files#diff-d4618f86870602ade7c1cb2d73453193R49 ?

**Comment by tymondesigns** [comment]
Another proposal - Should we reshuffle the folder structure slightly to better separate the tests from the src ? e.g.

```
> src
    > objects
    > components
    > etc
    - example.main.scss
> test
    - run.js
    - etc
- LICENSE
- CONTRIBUTING.md
- etc
```

**Comment by nenadjelovac** [comment]
@tymondesigns you tell me :) I have no testing experience :D. 

So in the given case (`@include inuit-font-size(21px, $important: true);`) expected result should be:

```sass
font-size: 1.3125rem !important;
line-height: 1.5 !important; //this is the default line height
```

Btw, your math is off, line height should be calculated as 1.5, not 1.6. Check [this SassMeister](http://www.sassmeister.com/gist/b82a9b01ef183c23ad41d07988533a50).

File: test/_tools.rem.scss [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+    $test: inuit-rem(18px);
+    $expect: 1.125rem;
+    @include assert-equal($test, $expect, "Returns the em equivalant for pixel value");        
+  }
+}
+
+@include test-module("[mixin] inuit-rem") {
+
+  @include test("it should output a property with an em value converted from a pixel value") {
```
Comment (nenadjelovac) [inline test/_tools.rem.scss]: Shouldn't it say `rem` instead of `em`.

**Comment by tymondesigns** [comment]
@nenadjelovac It was because I was passing in a line height value of `2.5` (all 3 params), and not explicitly specifying the `$important` param - [see here](https://github.com/inuitcss/inuitcss/pull/134/files#diff-d4618f86870602ade7c1cb2d73453193R49)

So the `line-height: 2.5` is the correct output in this case. Having said that, I think it *is* better to be explicit and pass in the parameter that is under test, so I will tweak that.

Hope that makes sense

**Comment by nenadjelovac** [comment]
@tymondesigns yeah, I got what you were trying with that test. I was just asking if there should be additional test with use case I explained in one of the previous comments?

**Comment by tymondesigns** [comment]
@nenadjelovac 👍 no worries... just updated that test case anyways

File: test/_tools.font-size.scss [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+      }
+
+      @include expect {
+        font-size: 1.125rem;
+        line-height: 2;
+      }
+    }
+  }
+
+  @include test("it should output a font-size in pixels when declaring as important") {
```
Comment (nenadjelovac) [inline test/_tools.font-size.scss]: The explanation is not correct. It should not output font-size in pixels. It should output `!important` as part of the value of `font-size` and `line-height` properties. Does it make sense? Sorry if I'm not explaining correctly.

Comment (tymondesigns) [inline test/_tools.font-size.scss]: hmm.. I can't see why that would output pixel values 😕 

to clarify we are talking about the following snippet yes ?

```sass
@include test("it should output a font-size and line-height when specifying a line-height") {

    @include assert() {

      @include output {
        @include inuit-font-size(18px, 2);
      }

      @include expect {
        font-size: 1.125rem;
        line-height: 2;
      }
    }
  }
```

Comment (tymondesigns) [inline test/_tools.font-size.scss]: oh I get it LOL

Comment (tymondesigns) [inline test/_tools.font-size.scss]: the description!

Sorry about that.. updated now

File: test/_tools.rem.scss (hunk 2 above)
Comment (tymondesigns) [inline test/_tools.rem.scss]: yes it should, good catch

**Comment by tymondesigns** [comment]
there is an issue with `sass-true` - it can't handle escaped class names it would seem. I have created an issue over there https://github.com/oddbird/true/issues/68

**Comment by anselmh** [comment]
> Another proposal - Should we reshuffle the folder structure slightly to better separate the tests from the src ? e.g.

This would make sense to me, yes. But it’s also out of scope for this PR and should be a separate one if we agree to do that change.

**Comment by anselmh** [comment]
Do you know which lib / tool calls the Sass compilation and where I can find this? I wonder how I could specify which sass version to use (libsass/rubysass)…

**Comment by anselmh** [comment]
Ah, I assume if true is fetched as ruby dependency it uses ruby-sass (not even that is true, it fetches compass as dependency in that case) and if fetched as js dependency, it uses node-sass? https://github.com/oddbird/true

**Comment by tymondesigns** [comment]
Circle comes with RVM right? so we just need to do something similar to the following for ruby I would think
``` bash
# configure ruby up here or define if circle config elsewhere?

> gem install sass true # install stuff
> {sass compile command here} # attempt to compile the sass - STDERR will notify on failure
> true-cli # run the tests via ruby
```

and potentially add the test config at `test/true.yml` (not sure if that's strictly necessary)

And after that run the node specific stuff. (we'll need to add `node-sass` as a dev dependency and add an npm script to compile the sass)

``` bash
> npm install
> node-sass example.main.scss -o dist # (this should be an npm script)  STDERR will notify on failure
> npm test
```

**Comment by anselmh** [comment]
Nice, that sounds good. Yes, Circle has rvm and nvm to manage ruby and node versions. I’ll try that later.

**Comment by anselmh** [comment]
So @tymondesigns, what mixins and functions have we covered already and what do we still need to cover?

**Comment by tymondesigns** [comment]
I would say that we shouldn't bother testing any mixins or functions that
do not have any parameters such as clearfix or the hidden mixin, as this
will offer no real benefit.

There is still an issue with widths testing as I described in #134 which
relates to a parsing error due to escaped characters, which I am yet to
properly dig into.

And I think that's it, unless I've missed any?

**Comment by anselmh** [comment]
Sounds good to me! 

So unless anyone of @inuitcss/core strongly objects here, I’d suggest to merge this now (since the width issue is an [external 4th-party blocker](https://github.com/mdevils/node-css-selector-parser/issues/13))?

**Comment by anselmh** [comment]
@tymondesigns Oh, forgot about this: Would you mind adding some notes on how to update the tests, and how to test things in https://github.com/inuitcss/inuitcss/blob/develop/CONTRIBUTING.md?

**Comment by tymondesigns** [comment]
Yea good point. I'll get on that today at some point

**Comment by anselmh** [comment]
Looks good. I think this is ready to merge then?

**Comment by tymondesigns** [comment]
oh shit.. how did I close that?!

**Comment by tymondesigns** [comment]
I'm done from my end 👍

**Comment by florianbouvot** [comment]
@anselmh it seems good but I'm not familiar with unit tests 😕

**Comment by anselmh** [comment]
As this works and adds a significant improvement to the previous workflow, I merged this now. Thanks again for the contribution, @tymondesigns — much appreciated!

**Comment by tymondesigns** [comment]
Awesome :) no worries

**Pull request merged**